telemetry/data_cluster_hardware.go

```diff
@@ -0,0 +1,155 @@
+	regexDiskAllowedNames         = regexp.MustCompile(`^(disk\d+s\d+|rootfs|devtmpfs|sd[a-z]\d*|vd[a-z]\d*|hd[a-z]\d*|nvme\d+(n\d+(p\d)?)?)$`)
+	sortedDiskAllowedPaths        = []string{"/", "/boot", "/dev", "/private/var/vm", "/System/Volumes/Data"}
+	sortedMemoryAllowedFieldNames = []string{"capacity"}
+)
+
+func init() {
+	sort.Strings(sortedCPUAllowedFieldNames)
+	sort.Strings(sortedDiskAllowedFieldNames)
+	sort.Strings(sortedDiskAllowedPaths)
+	sort.Strings(sortedMemoryAllowedFieldNames)
+}
+
+type clusterHardwareItem struct {
+	InstanceType   string                       `json:"instanceType"`
+	ListenHostHash string                       `json:"listenHostHash"`
+	ListenPort     string                       `json:"listenPort"`
+	CPU            map[string]string            `json:"cpu,omitempty"`
+	Memory         map[string]string            `json:"memory,omitempty"`
+	Disk           map[string]map[string]string `json:"disk,omitempty"`
+}
+
+func normalizeDiskName(name string) string {
+	const prefix = "/dev/"
+	if strings.HasPrefix(name, prefix) {
+		return name[len(prefix):]
+	}
+	return name
+}
+
+func isNormalizedDiskNameAllowed(name string) bool {
+	return regexDiskAllowedNames.MatchString(name)
+}
+
+func normalizeFieldName(name string) string {
+	return strcase.ToLowerCamel(name)
+}
+
+func getClusterHardware(ctx sessionctx.Context) ([]*clusterHardwareItem, error) {
+	rows, _, err := ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(`SELECT TYPE, INSTANCE, DEVICE_TYPE, DEVICE_NAME, NAME, VALUE FROM information_schema.cluster_hardware`)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// warning: the key of this variable is not hashed, it must not be returned directly
+	itemsByInstance := make(map[string]*clusterHardwareItem)
+L:
+	for _, row := range rows {
+		if row.Len() < 6 {
+			continue L
+		}
+		instance := row.GetString(1)
+		activeItem, ok := itemsByInstance[instance]
+		if !ok {
+			hostHash, port := parseAddressAndHash(instance)
+			activeItem = &clusterHardwareItem{
+				InstanceType:   row.GetString(0),
+				ListenHostHash: hostHash,
+				ListenPort:     port,
+				CPU:            make(map[string]string),
+				Memory:         make(map[string]string),
+				Disk:           make(map[string]map[string]string),
+			}
+			itemsByInstance[instance] = activeItem
+		}
+
+		deviceType := row.GetString(2)
+		deviceName := row.GetString(3)
+		fieldName := row.GetString(4)
+		fieldValue := row.GetString(5)
+
+		switch deviceType {
+		case "cpu":
+			if deviceName != "cpu" {
+				continue L
+			}
+			if !sortedStringContains(sortedCPUAllowedFieldNames, fieldName) {
+				continue L
+			}
+			activeItem.CPU[normalizeFieldName(fieldName)] = fieldValue
+		case "memory":
+			if deviceName != "memory" {
+				continue L
+			}
+			if !sortedStringContains(sortedMemoryAllowedFieldNames, fieldName) {
+				continue L
+			}
+			activeItem.Memory[normalizeFieldName(fieldName)] = fieldValue
+		case "disk":
```

Comment:
Do we need to check the value of `deviceName` in this case?

Reply:
deviceNames looks like sda, sdb, sdb1, etc. We have a whitelist of "safe" deviceNames that can be included in the telemetry data directory. For other names, they will be hashed (so that at least we can distinguish whether devices are the same when multiple components are deployed on the same host).

Reply:
Got it.